src/server/native_controllers/index.js

```diff
@@ -0,0 +1,88 @@
+import path from 'path';
+import { get } from 'lodash';
+import { Observable } from 'rxjs';
+import { spawnNativeController } from './spawn_native_controller';
+import { findPluginSpecs } from '../../plugin_discovery';
+import { safeChildProcess } from '../../utils/child_process';
+import { NativeController } from './native_controller';
+
+const getNativeControllers = async (settings) => {
+  const {
+    packageJson$,
+  } = findPluginSpecs(settings);
+
+  const spec$ = packageJson$
+    .mergeMap(packageJson => {
+      const nativeControllerSpecs = get(packageJson.contents, 'kibana.nativeControllers');
+      if (nativeControllerSpecs) {
+        return nativeControllerSpecs.map(spec => {
+          return {
+            pluginId: spec.pluginId,
+            path: path.resolve(packageJson.directoryPath, spec.path),
+            config: spec.config
+          };
+        });
+      }
+
+      return Observable.empty();
+    });
+
+  const nativeController$ = spec$
+    .mergeMap(async spec => {
+      const process = await spawnNativeController(settings, spec.path, spec.config);
+      return new NativeController(spec.pluginId, process);
+    });
+
+  return await nativeController$
+    .do(nativeController => {
+      const childProcessStatus = safeChildProcess(nativeController.process);
+      nativeController.process.on('exit', (code) => {
+        if (childProcessStatus.terminating || nativeController.killed) {
+          return;
+        }
+
+        console.error(`${nativeController.pluginId}'s native controller exited with code ${code}`);
+        process.exit();
```

Comment:
If we're going to forcefully exit the process (which I'm not 100% sure I like) we should absolutely be setting an `exitCode`.